lib/elixir/lib/kernel.ex

```diff
@@ -3094,43 +3137,62 @@ defmodule Kernel do
 
   defp in_range_literal(left, first, last) when first < last do
     quote do
-      :erlang.is_integer(unquote(left)) and
-        unquote(increasing_compare(left, first, last))
+      :erlang.andalso(:erlang.is_integer(unquote(left)),
+                      unquote(increasing_compare(left, first, last)))
     end
   end
 
   defp in_range_literal(left, first, last) do
     quote do
-      :erlang.is_integer(unquote(left)) and
-        unquote(decreasing_compare(left, first, last))
+      :erlang.andalso(:erlang.is_integer(unquote(left)),
+                      unquote(decreasing_compare(left, first, last)))
     end
   end
 
-  defp in_list(left, h, t) do
-    :lists.foldr(fn x, acc ->
-      quote do: :erlang.or(unquote(comp(left, x)), unquote(acc))
-    end, comp(left, h), t)
+  defp in_list(left, head, tail, expand, right, in_module?) do
+    :lists.foldr(fn elem, acc ->
+      quote do: :erlang.orelse(unquote(comp(left, elem, expand, right, in_module?)), unquote(acc))
+    end, comp(left, head, expand, right, in_module?), tail)
   end
 
-  defp comp(left, {:|, _, [h, t]}) do
-    quote(do: :erlang.or(:erlang."=:="(unquote(left), unquote(h)), unquote(left) in unquote(t)))
+  defp comp(left, {:|, _, [head, tail]}, expand, right, in_module?) do
+    case expand.(tail) do
+      [] ->
+        quote(do: :erlang."=:="(unquote(left), unquote(head)))
+      [tail_head | tail] ->
+        quote do
+          :erlang.orelse(:erlang."=:="(unquote(left), unquote(head)),
+                         unquote(in_list(left, tail_head, tail, expand, right, in_module?)))
+        end
+      tail when in_module? ->
+        quote do
+          :erlang.orelse(:erlang."=:="(unquote(left), unquote(head)),
+                         :lists.member(unquote(left), unquote(tail)))
+        end
+      _ ->
+        raise ArgumentError, <<"invalid args for operator \"in\", it expects a compile-time proper list ",
+                               "or compile-time range on the right side when used in guard expressions, got: ",
+                               Macro.to_string(right)::binary>>
+    end
   end
 
-  defp comp(left, right) do
+  defp comp(left, right, _expand, _right, _in_module?) do
     quote(do: :erlang."=:="(unquote(left), unquote(right)))
   end
 
   defp increasing_compare(var, first, last) do
     quote do
-      :erlang.">="(unquote(var), unquote(first)) and
-        :erlang."=<"(unquote(var), unquote(last))
+      :erlang.andalso(
```

Comment:
```
:erlang.andalso(:erlang.">="(unquote(var), unquote(first)),
                :erlang."=<"(unquote(var), unquote(last)))
```
This would be more consistent with the rest? (@whatyouhide has pointed it out the same in other places)